### PR TITLE
Remove `OwnedTableReference` and `OwnedSchemaReference`

### PR DESCRIPTION
**Pull request by comphead**
## Which issue does this PR close?

Related to #9764
Followup on #9824 

## Rationale for this change
In #9824 we switched to use `Arc` smart pointers instead of `CoW<&'a>`. Having done that the auxiliary types supporting lifetimes became obsolete. 

## What changes are included in this PR?
The PR removes `OwnedTableReference` and `OwnedSchemaReference` types which are now obsolete.

### CODE REVIEW

**Comment by comphead** [comment]
I have run benches, there is slight noise for most of benches, however there are sometimes regression 4%, sometimes performance improvement 4%.

My feeling once we unify `TableReference` types, we can also unify references to this types which will give more boost. Sometimes we use direct reference, sometime option, sometime lifetime, etc, unifying accessing to `TableReference` will make code cleaner and faster

**Review by phillipleblanc — COMMENTED** [review]
I like this change 👍. What is the process to mark this as a breaking change users should be aware of when updating?

**Comment by alamb** [comment]
> I like this change 👍. What is the process to mark this as a breaking change users should be aware of when updating?

We add a `api-change` label which then highlights it as a breaking change in the release notes https://github.com/apache/arrow-datafusion/blob/main/dev/changelog/36.0.0.md

File: datafusion/expr/src/logical_plan/builder.rs [hunk 1]
```diff
@@ -1152,36 +1152,35 @@ pub fn build_join_schema(
...
-                .map(|(q, f)| (q.map(|r| r.to_owned_reference()), f.clone()))
+                .map(|(q, f)| (q.cloned(), f.clone()))
                 .collect::<Vec<_>>();
             left_fields.into_iter().chain(right_fields).collect()
         }
         JoinType::Left => {
             // left then right, right set to nullable in case of not matched scenario
             let left_fields = left_fields
-                .map(|(q, f)| (q.map(|r| r.to_owned_reference()), f.clone()))
+                .map(|(q, f)| (q.cloned(), f.clone()))
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/builder.rs:1183]: 👍

**Review by alamb — APPROVED** [review]
This looks really nice -- thank you @comphead  🙏

File: datafusion/expr/src/logical_plan/builder.rs (hunk 1 above)
Comment (comphead) [reply datafusion/expr/src/logical_plan/builder.rs:1183]: I'm thinking to try  removing all this clones in future... if possible

**Comment by comphead** [comment]
I'll wait for couple of more hours, if there is no more comments, I'll merge to avoid conflicts on upcoming analyzer changes

**Pull request merged**